DBZ-453 support for newer PG wal2json with verbose type specifiers

rcoup opened the pull request:
Passes all unit tests with `mvn install -Pwal2json-decoder` both with PG built against the old https://github.com/eulerto/wal2json/tree/645ab69a & current https://github.com/eulerto/wal2json/tree/5352cc41 wal2json commits.

rcoup commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,105 +149,208 @@ public String getName() {
+        // leave up to the downstream message recipient to deal with.
         LOGGER.warn("processing column '{}' with unknown data type '{}' as byte array", columnName,
                 columnType);
         return rawValue.asBytes();
```
this path isn't covered by the tests. How to add?

rcoup replied:
a PostGIS geometry column should do it
eg. 
```
CREATE TABLE foo (id int primary key, geom geometry(POINT,4326))
INSERT INTO foo VALUES (1, 'SRID=4326;POINT(174 -36)'::geometry);
```

jpechane replied:
Here are gemotry types tested - https://github.com/debezium/debezium/blob/master/debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsStreamProducerIT.java#L104-L106

rcoup replied:
hmm, those are Postgresql's native "geometry" types, not the Postgis ones.

jpechane replied:
Yes, but you could extend the table with one of postgis types and check how it is processed

rcoup commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,105 +149,208 @@ public String getName() {
+        
+        if (rawValue.isNull()) {
+            // nulls are null
+            return null;
```
this path isn't covered by the tests. How to add?

jpechane replied:
Could you do it like int his method - https://github.com/debezium/debezium/blob/master/debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsStreamProducerIT.java#L71 just user an `INSERT` statement that inserts nulls?

rcoup commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,105 +149,208 @@ public String getName() {
+        String fullType = m.group("full");
+        String baseType = m.group("base").trim();
+        if (!Objects.toString(m.group("suffix"), "").isEmpty()) {
+            baseType = String.join(" ", baseType, m.group("suffix").trim());
```
this path isn't covered by the tests, I think a field using [`timestamp (0) with time zone`](https://www.postgresql.org/docs/9.6/static/datatype-datetime.html) should fix it. How to add?

jpechane replied:
You can extend the table here - https://github.com/debezium/debezium/blob/master/debezium-connector-postgres/src/test/resources/postgres_create_tables.ddl#L12

gunnarmorling commented:
Thanks a lot, @rcoup! We're going to take a look in depth next week.

@jpechane, I'm not quite sure were you're with leveraging that new column info we'll be getting in wal2json events. Is this related to it? Or is it even the same thing? Any refresher for my memory would be much appreciated :)

jpechane commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,105 +149,208 @@ public String getName() {
+            baseType = String.join(" ", baseType, m.group("suffix").trim());
+        }
+        String[] typeModifiers = Objects.toString(m.group("mod"), "").split(",\\s*", 0);  // TODO: use for decimal/etc precision
+        Boolean isArray = (m.group("array") != null);
```
Why do you use object `Boolean` instead of primitve type?

jpechane commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,105 +149,208 @@ public String getName() {
+        if (!Objects.toString(m.group("suffix"), "").isEmpty()) {
+            baseType = String.join(" ", baseType, m.group("suffix").trim());
+        }
+        String[] typeModifiers = Objects.toString(m.group("mod"), "").split(",\\s*", 0);  // TODO: use for decimal/etc precision
```
Woud it be possible to split it using precompiled regex?

jpechane commented:
Great work! Thanks. Could you please look at those few minor comments? Otherwise I think it is completed.

rcoup commented:
@jpechane I've added more tests (including for some custom pg-contrib types) and cleaned up the review comments.

But there's a couple of new test failures now that I think are related to DBZ-454:

* `RecordsStreamProducerIT.shouldReceiveChangesForInsertsWithDifferentDataTypes`
    `java.lang.AssertionError: Schema for org.apache.kafka.connect.data.Field@54a4c842 does not match the actual value expected:<Schema{STRING}> but was:<Schema{io.debezium.data.Json:STRING}>`

* `PostgresSchemaIT.shouldLoadSchemaForBuiltinPostgresTypes`:
    `java.lang.AssertionError: 'lt' has incorrect schema. expected:<Schema{STRING}> but was:<Schema{io.debezium.data.Json:STRING}>`

Any ideas?

jpechane commented:
Please rebase against the latest master. Regarding the issue - it seems the `ltree` type is somehow translated into JSON.

rcoup commented:
> Please rebase against the latest master.

Done

> Regarding the issue - it seems the ltree type is somehow translated into JSON.

The same thing was happening for Postgis types (DBZ-454). I think the default Java type somehow ends up as JSON? But I can't see where that happens...

jpechane commented:
Please check this
https://github.com/debezium/debezium/blob/57e7f84163dfa971095a4c23c008e1268c92eabd/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PgOid.java#L27
https://github.com/debezium/debezium/blob/57e7f84163dfa971095a4c23c008e1268c92eabd/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java#L89
`ltree` type is not known by Postgres OID class so a JDBC type is returned instead of it https://github.com/debezium/debezium/blob/57e7f84163dfa971095a4c23c008e1268c92eabd/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PgOid.java#L51
and it is the same value as oid for JSON

gunnarmorling commented:
Hey, I didn't look into this PR in detail, but got one quick question:

I was hoping that with this work we'd be able to get rid of that dreaded schema refresh via JDBC in `RecordStreamProducer#columnValues()`. It's still in there, though. Is this because we still need to apply the equivalent to this PR to the ProtoBuf path (in which case it'd be fine), or is there some other, more general reason why this has to remain?

jpechane commented:
As I mentioned in the past - we need to get done this and only after we can drop the schema refresh. I am now in parallel trying to build a soltuion to avoid schema refresh. When working we need to modify protobuf plugin too to get the information. But it is better fix wal2json and protobuf in the same PR.

gunnarmorling commented:
Ok, so your plan is to push your work on the ProtoBuf sid to this PR as well and as the last step - and within this PR - the mentioned schema refresh can finally be removed?

jpechane commented:
No the plan is to finish this PR to be able to work with extended wal2json types. And then I open another PR that will contain fix for DBZ-379.

gunnarmorling commented:
Ok, works for me. Last question then: in the comments inline it says wal2json will be supported before and after that commit with the additional type info. What will happen when using the old wal2json plug-in (e.g. in RDS before it gets updated there) and we don't have the JDBC schema refresh any more?

gunnarmorling commented:
Excellent work btw. @rcoup, that's really much appreciated and I'm excited about this functionality :)

jpechane commented:
I like the way how to deal with unknow types. Please rebase again to master and I've found one more issue in the code. Other than that I am happy with it.

jpechane commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,107 +150,210 @@ public String getName() {
+        if (!Objects.toString(m.group("suffix"), "").isEmpty()) {
+            baseType = String.join(" ", baseType, m.group("suffix").trim());
+        }
+        String[] typeModifiers = TYPEMOD_PATTERN.split(Objects.toString(m.group("mod"), ""));  // TODO: use for decimal/etc precision
```
This code returns array of size 1 even when there are no modifiers. In such a case there should be an array of size 0.

rcoup commented:
All done :)

jpechane commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,110 +150,213 @@ public String getName() {
+        if (!Objects.toString(m.group("suffix"), "").isEmpty()) {
+            baseType = String.join(" ", baseType, m.group("suffix").trim());
+        }
+        String[] typeModifiers = {};
```
Will this create a new array instance for each invocation or will be there only one shared?

rcoup replied:
_`<not a java expert>`_ 
AFAIK the empty array is static & shared. If there are any modifiers found, then the `typeModifiers` reference is pointed to the new array from the regex 
_`</not a java expert>`_

jpechane replied:
I tried this
```java
public class Test {

    @org.junit.Test
    public void test() {
        emptyArray();
        emptyArray();
        emptyArray();
    }

    private void emptyArray() {
        final String[] typeModifiers = {};
        System.out.println(typeModifiers);
    }
}
```
and got a unique value for each call

rcoup replied:
So... how do I fix it?

jpechane replied:
I think it is goin to be easy - just declare `private static final String[]` and use it as the empty array returned.

jpechane commented:
I've added one last comment - otherwise I am fine with it. @gunnarmorling Do you have any comments on the PR?

gunnarmorling commented on debezium-connector-postgres/src/test/resources/postgres_create_tables.ddl:
```diff
@@ -3,18 +3,22 @@
 -- Generate a number of tables to cover as many of the PG types as possible
 DROP SCHEMA IF EXISTS public CASCADE;
 CREATE SCHEMA public;
+CREATE EXTENSION IF NOT EXISTS ltree SCHEMA public;
```
Could you add a comment what these are used for?

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,110 +150,213 @@ public String getName() {
             case "numeric":
-                return rawValue.isNotNull() ? rawValue.asDouble() : null;
+            case "decimal":
+                // TODO: When wal2json supports decimal/numeric, return via .asBigDecimal()
```
That seems to be the case now?

rcoup replied:
Yeah, it does, I verified the toString() representation is "right", but I haven't figured out the JSON-parsing side in Java yet.

See numeric comments above.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,110 +150,213 @@ public String getName() {
+
+        switch (baseType) {
+            // include all types from https://www.postgresql.org/docs/current/static/datatype.html#DATATYPE-TABLE
+            // plus aliases from the shorter names produced by older wal2json
```
Could we group the old names and move them all towards the end? That way we can git rid of them more easily in the future, once we say we don't support them any longer.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,110 +150,213 @@ public String getName() {
+
+            // catch-all for other known/builtin PG types
+            // TODO: improve with more specific/useful classes here?
+            case "bit":
```
By listing them here the log message below won't be printed. I'd say it should be enough to have just the `default` clause and then leave the method with the `return` below.

rcoup replied:
these are types specifically builtin to PG which is why I listed them, so seems we should list them. But can fall-through to default easily enough.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,110 +150,213 @@ public String getName() {
                 }
-                catch (SQLException e) {
-                    LOGGER.warn("Unexpected exception trying to process PgArray column '{}'", columnName, e);
+            case "polygon":
```
Do we have a test somewhere for these geometric types? I don't think they where supported via DecoderBufs (with the exception of `POINT`).

rcoup replied:
They're builtin types, but I think there's only a test for point. Not sure if they're used much/at-all, IME (as a been-doing-this-way-too-long geo developer) everyone uses PostGIS :)

rcoup commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,110 +150,213 @@ public String getName() {
+        }
+        String[] typeModifiers = {};
+        if (m.group("mod") != null) {
+            typeModifiers = TYPEMOD_PATTERN.split(m.group("mod"));  // TODO: use for decimal/etc precision
```
When @jpechane and I discussed, it seemed he had plans for type modifiers. I think parsing of numerics/datetimes/etc from raw JSON depends on precision as we want to do it properly.

eg. NUMERIC(5,2) will be represented (correctly) as `123.45` in the _raw_ JSON, but by default we get it in java as a double.

If we know it's meant to be a numeric, we should be passing it as a Decimal through, with the right precision. But I felt that was out of scope, so did @jpechane.

Compiler will optimise out in the meantime :)

rcoup commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/Wal2JsonReplicationMessage.java:
```diff
@@ -139,110 +150,213 @@ public String getName() {
+                final String dataString = rawValue.asString();
+                PgArray arrayData = new PgArray(connection.get(), connection.get().getTypeInfo().getPGArrayType(fullType), dataString);
+                Object deserializedArray = arrayData.getArray();
+                // TODO: what types are these? Shouldn't they pass through this function again?
```
I don't know enough about it.

rcoup commented:
Addressed more comments :)

rcoup commented:
Pushed again with static array fix. I fixed the tests too. ✅

gunnarmorling commented:
Pushed a few changes to this branch as described above. I've also removed the obsolete code for dealing with the type modifiers. Let's re-instantiate it for building the schema once we get there.

gunnarmorling commented:
Rebased and applied. Thanks, @rcoup!
